### PR TITLE
cstor-tracking#53 destroy pool fails with EBUSY if there are zvols on it

### PR DESCRIPTION
**Pull request by gila**
This PR addresses two issues observed during pool export and during destroy. 

The first problem was that when during pool import and export, we iterate over a certain set of function pointers that do post import and pre-export tasks. It was observed that when there where no volumes created on the pool the export process sometimes, appeared to be hanging. 

There where two threads running that should have been shut down during the post-export/destroy process as can be examined from the core: 

```
(lldb) print *(kthread_t*)0x0000000000c85530
(kthread_t) $9 = {
  t_tid = 140737224869632
  t_func = 0x00007ffff6f363f0 (libzpool.so.2`uzfs_update_txg_zap_thread at uzfs_zap.c:133)
  t_arg = 0x0000000000c601f0
  t_pri = 0
}
```
And this thread:

```
* thread #1, name = 'lt-test_export', stop reason = signal SIGSTOP
    frame #0: libpthread.so.0`__GI___nanosleep(requested_time=0x00007fffffffe010, remaining=0x0000000000000000) at nanosleep.c:27
  * frame #1: libzpool.so.2`close_txg_update_thread(s=0x0000000000c601f0) at uzfs_task.c:93
(lldb) l uzfs_task.c:93
   83          struct timespec ts;
   84
   85          mutex_enter(&(uzfs_spa(spa)->mtx));
   86          cv_signal(&(uzfs_spa(spa)->cv));
   87          mutex_exit(&(uzfs_spa(spa)->mtx));
   88
   89          ts.tv_sec = 0;
   90          ts.tv_nsec = 100000000;
   91
   92          while (uzfs_spa(spa)->update_txg_tid != NULL)
   93              nanosleep(&ts, NULL);
(lldb)
```

It will only exit once `update_txg_tid` == NULL which will be set by `uzfs_update_txg_zap_thread` which is also waiting on a mutex.  Printing the mutex, showed that the owner field is empty 
```
__owner = 0,
```

Which means that the call to cv_timedwait() has missed the signal during the pre-export phase which causes us to wait for 60 seconds (and thus not really hang)

The second issue, when there *are* zvols during the opening of the pool we add a tag to the SPA, we did not release this tag during pool export, which prevented the pool to hang and wait forever. 

Lastly, the `zvol_list_mutex` was of type `pthread_t` which should have been of type `kmutex_t` (which trips an assertion) and is now initialized during uzfs_init() note that the later issue was only visible during debug builds. 

To verify all of this, new test cases have been added that I think, cover all the cases we are interested in. 

Signed-off-by: Jeffry Molanus <jeffry.molanus@openebs.io>

### CODE REVIEW

**Review by mynktl — APPROVED** [review]
looks good to me.

File: lib/libzpool/uzfs_zap.c [hunk 1]
```diff
@@ -133,8 +133,9 @@ uzfs_update_txg_zap_thread(void *s)
 {
 	spa_t *spa = (spa_t *)s;
 	uint64_t txg = spa_last_synced_txg(spa);
-
 	mutex_enter(&(uzfs_spa(spa)->mtx));
+	if (uzfs_spa(spa)->close_pool == 1)
```
Comment (vishnuitta) [inline lib/libzpool/uzfs_zap.c]: why do we need this change?

File: module/zfs/spa.c [hunk 2]
```diff
@@ -4570,6 +4570,7 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
 	if ((new_state == POOL_STATE_DESTROYED) ||
 	    (new_state == POOL_STATE_EXPORTED)) {
 		uzfs_zvol_destroy_cb(spa_name(spa), NULL);
+		uzfs_close_pool(spa);
```
Comment (vishnuitta) [inline module/zfs/spa.c]: is this related to 'uzfs_open_pool' in zrepl_import? does this mean - if any pool/volume created from CLI will get exported and destroyed?

Comment (vishnuitta) [reply module/zfs/spa.c]: I mean - if I understood correctly, without this uzfs_close_pool, if any pool/volume created from CLI will get exported and destroyed.

File: tests/cbtest/gtest/test_export.cc [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+/*
+ * if you find the need to copy paste the following
+ * two functions consider to put them in a generic
+ * header file, because I already copied them =)
+ *
+ */
+
+std::string getCmdPath(std::string zfsCmd) {
```
Comment (vishnuitta) [inline tests/cbtest/gtest/test_export.cc]: there seems to be this functions already existing. It would be better to reuse them.

Comment (vishnuitta) [reply tests/cbtest/gtest/test_export.cc]: these* functions

File: tests/cbtest/gtest/test_export.cc [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+	CreatePool() : zi(NULL), zv(NULL), spa(NULL) {}
+
+	virtual void SetUp() override {
+		uzfs_init();
+		libuzfs_ioctl_init();
+	}
+
+
+	void createPool() {
+		ASSERT_TRUE(uzfs_create_pool(pool_name, DISK, &spa ) == 0);
```
Comment (vishnuitta) [inline tests/cbtest/gtest/test_export.cc]: this PR being to fix CLI, its better to create pool/datasets through CLI, by having 'zrepl' running

File: lib/libzpool/uzfs_zap.c (hunk 1 above)
Comment (vishnuitta) [reply lib/libzpool/uzfs_zap.c]: ok.. got to see the need of this..

File: tests/cbtest/gtest/test_export.cc (hunk 3 above)
Comment (gila) [reply tests/cbtest/gtest/test_export.cc]: They where replaced very recently by Jan.. for future references yes, we should have a cb_common.hpp or something

File: module/zfs/spa.c (hunk 2 above)
Comment (gila) [reply module/zfs/spa.c]: We need to open pool uzfs_open_pool() and we need to close them using uzfs_close_pool() if we dont close it here, we keep the UZFS_SPA_TAG held.

File: tests/cbtest/gtest/test_export.cc (hunk 4 above)
Comment (gila) [reply tests/cbtest/gtest/test_export.cc]: These tests pool related feature without zrepl -- and we should not depend on zrepl as it "just an application" for running any test cases against core libzpool features.

Comment (gila) [reply tests/cbtest/gtest/test_export.cc]: This PR tests CLI and API functions BTW.

File: module/zfs/spa.c (hunk 2 above)
Comment (gila) [reply module/zfs/spa.c]: This is needed because our pool open has pre/post hooks and puts a hold on the spa

Comment (vishnuitta) [reply module/zfs/spa.c]: I want to check whether all the cases of pool export / destroy is fixed.
Two cases that I'm aware are:
1. import pool using 'zrepl import -t <ip>'
2. created pool using 'zpool create ..' after starting zrepl as 'zrepl start -t <ip>'
I understand change is needed.. :) I can see where case 1 being fixed.
Does case 2 also have issue and got fixed with this? or case 2 was already working?

File: tests/cbtest/gtest/test_export.cc (hunk 3 above)
Comment (vishnuitta) [reply tests/cbtest/gtest/test_export.cc]: ok.. makes sense

File: tests/cbtest/gtest/test_export.cc (hunk 4 above)
Comment (vishnuitta) [reply tests/cbtest/gtest/test_export.cc]: hmm.. I see CLI for zpool destroy/export,  and zfs create. But, pool is being created and imported using APIs which won't be the case usually.
And, code is duplicated for import, which actually tests unit test code rather than production code.

File: include/zrepl_mgmt.h [hunk 5]
```diff
@@ -36,7 +37,7 @@ extern "C" {
 #define	uZFS_ZVOL_WORKERS_MAX 128
 #define	uZFS_ZVOL_WORKERS_DEFAULT 6
 
-extern pthread_mutex_t zvol_list_mutex;
```
Comment (jkryl) [inline include/zrepl_mgmt.h:40]: this should be declared with extern and the actual definition should be in uzfs_mgmt.c. Otherwise every file which includes this header file will have a global variable of this name.

File: tests/cbtest/gtest/test_export.cc [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+	}
+
+	void exportPool() {
+		uzfs_close_pool(spa);
+		ASSERT_EQ(spa_export(pool_name, NULL, B_TRUE, B_FALSE), 0);
+	}
+
+
+	void finish() {
```
Comment (jkryl) [inline tests/cbtest/gtest/test_export.cc:142]: just a nit: this could be renamed to TearDown() - a hook called automatically at the end of test. That would nicely match with with SetUp().

**Review by jkryl — COMMENTED** [review]
nice fix

File: include/zrepl_mgmt.h (hunk 5 above)
Comment (gila) [reply include/zrepl_mgmt.h:40]: thx jan fixed

**Review by jkryl — APPROVED** [review]
